core/internal/runfiles/saved_file.go

```diff
@@ -106,6 +111,26 @@ func (f *savedFile) Upload(
 //
 // It must be called while a lock is held. It temporarily releases the lock.
 func (f *savedFile) doUpload(uploadURL string, uploadHeaders []string) {
+	// Mark as uploading first so concurrent f.Upload calls defer and queue.
+	f.isUploading = true
+
+	// Check if the file bytes differ from the last successful upload.
+	f.Unlock()
+	currentB64MD5, err := hashencode.ComputeFileB64MD5(f.realPath)
+	if err != nil {
+		currentB64MD5 = "" // treat as "changed" below
+	}
+	f.Lock()
+	if currentB64MD5 != "" && f.lastUploadedB64MD5 == currentB64MD5 {
+		// No changes: clear in-flight state and honor any queued reupload.
```

Comment:
@timoffex since we didn't schedule a task in the first place (and didn't do `f.wg.Add(1)`), I think doing this here explicitly is better.